### PR TITLE
Add kills left to xp tracker and globe

### LINKED ISSUE #1594: Add "Kills Left" on the XP Tracker for combat kills
Right now the combat XP Tracker shows "### Action Left" until a level up.  If the client could be updated with "### Kills Left" for the combat skills that would help out a lot.  Because right now I am 62 ranged, and I look at the XP tracker and it shows "1612 Actions Left" and than a few second later "1323 Actions Left", few seconds after that "1852 Actions Left".  It is not very accurate when it comes to combat skills.

### PR DESCRIPTION
**Pull request by DaanG96**
I closed #1645 and implemented the feature after a few changes were made in the xp tracker code. Also added the kills left to the xp globe now.

screenshot
closes #1594

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/xpglobes/XpGlobesOverlay.java [hunk 1]
```diff
@@ -282,4 +281,11 @@ private void drawTooltipIfMouseover(Graphics2D graphics, XpGlobe mouseOverSkill,
 		xpTooltip.render(graphics);
 		graphics.translate(-x, -y);
 	}
+
+	private String getActionKillsText(String skillName)
+	{
+		String[] combatSkills = new String[]{"Attack", "Defence", "Hitpoints", "Ranged", "Strength"};
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xpglobes/XpGlobesOverlay.java]: Instead of using raw strings, use the Skill enum

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 2]
```diff
@@ -74,9 +82,17 @@
 	private final JLabel expHour = new JLabel();
 	private final JLabel expLeft = new JLabel();
 	private final JLabel actionsLeft = new JLabel();
+	private Map<String, Integer> oppInfoHealth = OpponentInfoPlugin.loadNpcHealth();
+	@Getter(AccessLevel.PACKAGE)
+	private static final Skill[] COMBAT = new Skill[]
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: Declare this as list, move this next to other static final variables

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 3]
```diff
@@ -156,6 +172,8 @@
 		progressBar.setComponentPopupMenu(popupMenu);
 
 		add(container, BorderLayout.NORTH);
+
+		killsRemaining = Integer.MAX_VALUE;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: Set this to Integer.MAX_VALUE when declaring the variable

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 4]
```diff
@@ -183,7 +201,14 @@ private void rebuildAsync(boolean updated, XpSnapshotSingle xpSnapshotSingle)
 			// Update information labels
 			expGained.setText(htmlLabel("XP Gained: ", xpSnapshotSingle.getXpGainedInSession()));
 			expLeft.setText(htmlLabel("XP Left: ", xpSnapshotSingle.getXpRemainingToGoal()));
-			actionsLeft.setText(htmlLabel("Actions: ", xpSnapshotSingle.getActionsRemainingToGoal()));
+			if (Arrays.asList(COMBAT).contains(skill))
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: Dont use arraysAsList every time, use lists as mentioned earlier. Also, put new empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 5]
```diff
@@ -183,7 +201,14 @@ private void rebuildAsync(boolean updated, XpSnapshotSingle xpSnapshotSingle)
...
+			{
+				actionsLeft.setText(htmlLabel("Kills: ", getKillsRemaining(xpSnapshotSingle)));
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: Use ternary operator

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerServiceImpl.java [hunk 6]
```diff
@@ -54,6 +56,15 @@ public int getActionsHr(Skill skill)
 	@Override
 	public int getActionsLeft(Skill skill)
 	{
+		Skill[] COMBAT = XpInfoBox.getCOMBAT();
+		if (Arrays.asList(COMBAT).contains(skill))
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerServiceImpl.java]: Why? This entire method looks incorrect.

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 7]
```diff
@@ -206,6 +231,34 @@ private void rebuildAsync(boolean updated, XpSnapshotSingle xpSnapshotSingle)
...
+	{
+		Actor opponent = client.getLocalPlayer().getInteracting();
+		if (opponent != null)
+		{
+			int opponentHealth = oppInfoHealth.get(opponent.getName() + "_" + opponent.getCombatLevel());
+			double modifier = getCombatXPModifier(skill);
+			double actionExp = (opponentHealth * modifier);
+			killsRemaining = (int) Math.ceil(xpSnapshotSingle.getXpRemainingToGoal() / actionExp);
+		}
+		return killsRemaining;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 8]
```diff
@@ -206,6 +231,34 @@ private void rebuildAsync(boolean updated, XpSnapshotSingle xpSnapshotSingle)
 		expHour.setText(htmlLabel("XP/Hour: ", xpSnapshotSingle.getXpPerHour()));
 	}
 
+	int getKillsRemaining(XpSnapshotSingle xpSnapshotSingle)
+	{
+		Actor opponent = client.getLocalPlayer().getInteracting();
+		if (opponent != null)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 9]
```diff
@@ -206,6 +231,34 @@ private void rebuildAsync(boolean updated, XpSnapshotSingle xpSnapshotSingle)
...
+			double modifier = getCombatXPModifier(skill);
+			double actionExp = (opponentHealth * modifier);
+			killsRemaining = (int) Math.ceil(xpSnapshotSingle.getXpRemainingToGoal() / actionExp);
+		}
+		return killsRemaining;
+	}
+
+	private double getCombatXPModifier(Skill skill)
+	{
+		final double sharedXPModifier = 4.0 / 3.0;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: Constants maybe?

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 10]
```diff
@@ -206,6 +231,34 @@ private void rebuildAsync(boolean updated, XpSnapshotSingle xpSnapshotSingle)
...
+		final double longRangedXPModifier = 2.0;
+		final double defaultModifier = 4;
+		if (skill.equals(Skill.HITPOINTS))
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 11]
```diff
@@ -206,6 +231,34 @@ private void rebuildAsync(boolean updated, XpSnapshotSingle xpSnapshotSingle)
...
+		{
+			return sharedXPModifier;
+		}
+		int styleIndex = client.getVar(VarPlayer.ATTACK_STYLE);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 12]
```diff
@@ -206,6 +231,34 @@ private void rebuildAsync(boolean updated, XpSnapshotSingle xpSnapshotSingle)
...
+		WeaponType weaponType = WeaponType.getWeaponType(client.getVar(Varbits.EQUIPPED_WEAPON_TYPE));
+		return weaponType.getAttackStyles()[styleIndex].equals(AttackStyle.CONTROLLED) ? sharedXPModifier :
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: Properly format this line , eg
check
  ?
  :

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpPanel.java [hunk 13]
```diff
@@ -53,6 +55,7 @@
 @Slf4j
 class XpPanel extends PluginPanel
 {
+	@Getter(AccessLevel.PACKAGE)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpPanel.java]: This looks incorrect

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerPlugin.java [hunk 14]
```diff
@@ -75,6 +77,7 @@
 	private ScheduledExecutorService executor;
 
 	private NavigationButton navButton;
+	@Getter(AccessLevel.PACKAGE)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerPlugin.java]: Dont expose everything. Rather, try to figure out how to do this with current access levels

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 15]
```diff
@@ -74,9 +83,16 @@
 	private final JLabel expHour = new JLabel();
 	private final JLabel expLeft = new JLabel();
 	private final JLabel actionsLeft = new JLabel();
+	private final Map<String, Integer> oppInfoHealth = OpponentInfoPlugin.loadNpcHealth();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: This should add a dependency on the opponent info plugin probably.. or we move the opponent health stuff into the core client somewhere.

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 16]
```diff
@@ -206,6 +225,42 @@ private void rebuildAsync(boolean updated, XpSnapshotSingle xpSnapshotSingle)
 		expHour.setText(htmlLabel("XP/Hour: ", xpSnapshotSingle.getXpPerHour()));
 	}
 
+	int getKillsRemaining(XpSnapshotSingle xpSnapshotSingle)
+	{
+		Actor opponent = client.getLocalPlayer().getInteracting();
+		
+		if (opponent != null)
+		{
+			int opponentHealth = oppInfoHealth.get(opponent.getName() + "_" + opponent.getCombatLevel());
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: This will crash if there is no hp for the opponent

File: runelite-client/src/main/java/net/runelite/client/plugins/xpglobes/XpGlobesOverlay.java [hunk 17]
```diff
@@ -208,6 +210,7 @@ private void drawSkillImage(Graphics2D graphics, XpGlobe xpGlobe, int x, int y)
 
 	private void drawTooltipIfMouseover(Graphics2D graphics, XpGlobe mouseOverSkill, Ellipse2D drawnGlobe)
 	{
+		List<Skill> combat = Arrays.asList(Skill.ATTACK, Skill.STRENGTH, Skill.DEFENCE, Skill.RANGED, Skill.HITPOINTS);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/xpglobes/XpGlobesOverlay.java]: `private static final` and `ImmutableList`

File: runelite-client/src/main/java/net/runelite/client/plugins/xpglobes/XpGlobesOverlay.java [hunk 18]
```diff
@@ -282,4 +285,14 @@ private void drawTooltipIfMouseover(Graphics2D graphics, XpGlobe mouseOverSkill,
 		xpTooltip.render(graphics);
 		graphics.translate(-x, -y);
 	}
+
+	private String getActionKillsText(String skillName)
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xpglobes/XpGlobesOverlay.java]: Combat skills can get experience with non-kill actions too, though rare.

File: runelite-client/src/main/java/net/runelite/client/plugins/xpglobes/XpGlobesOverlay.java [hunk 19]
```diff
@@ -282,4 +285,14 @@ private void drawTooltipIfMouseover(Graphics2D graphics, XpGlobe mouseOverSkill,
...
+	{
+		List<Skill> COMBAT = Arrays.asList(Skill.ATTACK, Skill.STRENGTH, Skill.DEFENCE, Skill.RANGED, Skill.HITPOINTS);
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xpglobes/XpGlobesOverlay.java]: Don't we have a combat skills list somewhere already?
Like before this PR was made

File: runelite-client/src/main/java/net/runelite/client/plugins/xpglobes/XpGlobesOverlay.java [hunk 20]
```diff
@@ -282,4 +285,14 @@ private void drawTooltipIfMouseover(Graphics2D graphics, XpGlobe mouseOverSkill,
...
+		return COMBAT.stream()
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xpglobes/XpGlobesOverlay.java]: Although concise, this is overdoing it.. 

You might as well just use an ImmutableSet then do .contains(skill) to check.

On that note, can you not use skillName and actually use the skill enum? It looks like you may have to do some data passing to get it.

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 21]
```diff
@@ -206,6 +225,42 @@ private void rebuildAsync(boolean updated, XpSnapshotSingle xpSnapshotSingle)
...
+			double modifier = getCombatXPModifier(skill);
+			double actionExp = (opponentHealth * modifier);
+			killsRemaining = (int) Math.ceil(xpSnapshotSingle.getXpRemainingToGoal() / actionExp);
+		}
+
+		return killsRemaining;
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: So, if it is a combat skill, and the player used a lamp after logging in, they will see MAX_INT kills remaining?
And if they fought, but then used a lamp, their killsRemaining will be the same.

Try to fall back to actions remaining (and say actions instead of kills) when no opponent can be determined.

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 22]
```diff
@@ -206,6 +225,42 @@ private void rebuildAsync(boolean updated, XpSnapshotSingle xpSnapshotSingle)
...
+
+		if (skill.equals(Skill.HITPOINTS))
+		{
+			return sharedXPModifier;
+		}
+
+		int styleIndex = client.getVar(VarPlayer.ATTACK_STYLE);
+		WeaponType weaponType = WeaponType.getWeaponType(client.getVar(Varbits.EQUIPPED_WEAPON_TYPE));
+
+		return weaponType.getAttackStyles()[styleIndex].equals(AttackStyle.CONTROLLED)
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: Multi-ternary expressions are hard to follow.

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpPanel.java [hunk 23]
```diff
@@ -53,6 +54,7 @@
 @Slf4j
 class XpPanel extends PluginPanel
 {
+	@Getter
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpPanel.java]: Plz don't

This is already messy as is, don't hook into other classes internal state so directly.

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerService.java [hunk 24]
```diff
@@ -49,6 +49,10 @@
 	 */
 	int getActionsLeft(Skill skill);
 
+	/**
+	 *
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerService.java]: dis javadoc needs help.

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerServiceImpl.java [hunk 25]
```diff
@@ -57,6 +57,12 @@ public int getActionsLeft(Skill skill)
 		return plugin.getSkillSnapshot(skill).getActionsRemainingToGoal();
 	}
 
+	@Override
+	public int getKillsLeft(Skill skill)
+	{
+		return plugin.getXpPanel().getInfoBoxes().get(skill).getKillsRemaining(plugin.getSkillSnapshot(skill));
```
Comment (LeviSchuck) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerServiceImpl.java]: Ugh, just make a function on the plugin to actually get kills remaining instead of throwing this crapton of nesting getters here. What if any of them is null in the future? This is fragile

**Comment by DaanG96** [comment]
IRL is taking over so it will be a while before I will get back on this. It needs more work for sure, so it will take some time

**Comment by psikoi** [comment]
Status on this?

**Comment by LeviSchuck** [comment]
I can probably wake up and take this on.

**Comment by MagicfTail** [comment]
Added in #6163 and #6560
